README.md
# egg-rpc

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/egg-rpc-base.svg?style=flat-square
[npm-url]: https://npmjs.org/package/egg-rpc-base
[travis-image]: https://img.shields.io/travis/eggjs/egg-rpc.svg?style=flat-square
[travis-url]: https://travis-ci.org/eggjs/egg-rpc
[codecov-image]: https://img.shields.io/codecov/c/github/eggjs/egg-rpc.svg?style=flat-square
[codecov-url]: https://codecov.io/github/eggjs/egg-rpc?branch=master
[david-image]: https://img.shields.io/david/eggjs/egg-rpc.svg?style=flat-square
[david-url]: https://david-dm.org/eggjs/egg-rpc
[snyk-image]: https://snyk.io/test/npm/egg-rpc-base/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/egg-rpc-base
[download-image]: https://img.shields.io/npm/dm/egg-rpc-base.svg?style=flat-square
[download-url]: https://npmjs.org/package/egg-rpc-base

egg-rpc plugin for egg framework

## Install

```bash
$ npm i egg-rpc-base --save
```

## Usage

### Enable the plugin

Change `${app_root}/config/plugin.js` to enable egg-rpc plugin:

```js
exports.rpc = {
  enable: true,
  package: 'egg-rpc-base',
};
```

### RPC Service Discovery

By default, We use zookeeper for service discovery. Therefore you need to configure the zk address:

```js
// ${app_root}/config/config.${env}.js
config.rpc = {
  registry: {
    address: '127.0.0.1:2181', // configure your real zk address
  },
};
```

We plan to provide more implementations of service discovery. And also you can implement it by yourself, you can follow this [article](https://github.com/eggjs/egg-sofa-rpc/wiki/%E8%87%AA%E5%AE%9A%E4%B9%89%E6%9C%8D%E5%8A%A1%E5%8F%91%E7%8E%B0%E5%AE%9E%E7%8E%B0)

### RPC Client

By using the plugin, you can call rpc services provided by other system.

#### 1. Get the RPC Interface Definition

We use [protobuf interface definition lanaguge](https://developers.google.com/protocol-buffers/docs/proto3) to describe the RPC service. So you need to get the \*.proto files and put them into `${app_root}/proto` folder

#### 2. Global RPC Client Configuration

Configure RPC Client information in ${app_root}/config/config.{env}.js:

```js
// ${app_root}config/config.${env}.js
exports.rpc = {
  client: {
    responseTimeout: 3000,
  },
};
```

- `responseTimeout`(optional): RPC timeout in milliseconds, default value is 3000

#### 3. Configure the Interface in proxy.js

`${app_root}/config/proxy.js` is a very important config file for rpc client, you should configure the services you needed, then executing the `egg-rpc-generator` tool to generate the proxy files.

Let's see a simple example of proxy.js. It declare a interface named: `org.eggjs.rpc.test.ProtoService` provided by `rpc-demo` application

```js
'use strict';

module.exports = {
  services: [{
    appName: 'rpc-demo',
    api: {
      ProtoService: 'org.eggjs.rpc.test.ProtoService',
    },
  }],
};
```

Refer this [acticle](https://github.com/eggjs/egg-sofa-rpc/wiki/RPC-%E4%BB%A3%E7%90%86%EF%BC%88Proxy%EF%BC%89%E9%85%8D%E7%BD%AE) for more details

#### 4. Generate the Proxy

Run `egg-rpc-generator` to generate the proxy files. After running success, it will generate a file: `ProtoService.js` under `${app_root}/app/proxy`

```bash
$ egg-rpc-generator

[EggRpcGenerator] framework: /egg-rpc-demo/node_modules/egg, baseDir: /egg-rpc-demo
[ProtoRPCPlugin] found "org.eggjs.rpc.test.ProtoService" in proto file
[ProtoRPCPlugin] save all proto info into "/egg-rpc-demo/run/proto.json"
```

#### 5. Call the Service

You can call the RPC service by using `ctx.proxy.proxyName`. The proxyName is `key` value of api object you configure in proxy.js. In our example, it's ProtoService, and proxyName using lower camelcase, so it's `ctx.proxy.protoService`

```js
'use strict';

const Controller = require('egg').Controller;

class HomeController extends Controller {
  async index() {
    const { ctx } = this;
    const res = await ctx.proxy.protoService.echoObj({
      name: 'gxcsoccer',
	    group: 'A',
    });
    ctx.body = res;
  }
}

module.exports = HomeController;
```

As above, you can call remote service as a local method.

### RPC Server

By using the plugin, you also can publish your own RPC service to other system.

#### 1. Define the RPC Interface

Writing the \*.proto files, and put them into `${app_root}/proto` folder

```
# ProtoService.proto
syntax = "proto3";

package org.eggjs.rpc.test;
option java_multiple_files = true; // 可选
option java_outer_classname = "ProtoServiceModels"; // 可选

service ProtoService {
    rpc echoObj (EchoRequest) returns (EchoResponse) {}
}

message EchoRequest {
    string name = 1;
    Group group = 2;
}

message EchoResponse {
    int32 code = 1;
    string message = 2;
}

enum Group {
    A = 0;
    B = 1;
}
```

#### 2. Global RPC Server Configuration

Configure RPC Server information in ${app_root}/config/config.{env}.js:

```js
module.exports = {
	rpc: {
    server: {
      namespace: 'org.eggjs.rpc.test',
    },
	},
},
```

- `namespace`(required): the default namespace of all rpc service
- `selfPublish`(optional): whether every node process publish service independent, default is true
- `port`(optional): the TCP port will be listen on
- `maxIdleTime`(optional): server will disconnect the socket if idle for long time
- `responseTimeout`(optional): Number of milliseconds to wait for a response to begin arriving back from the remote system after sending a request
- `codecType`(optional): recommended serialization method，default is protobuf

#### 3. Implemenation the RPC Interface

Put your implementation code under `${app_root}/app/rpc` folder

```js
// ${app_root}/app/rpc/ProtoService.js
exports.echoObj = async function(req) {
  return {
    code: 200,
    message: 'hello ' + req.name + ', you are in ' + req.group,
  };
};
```

#### 4. RPC Unittest in Eggjs

RPC Client Unittest (mock)

```js
'use strict';

const mm = require('egg-mock');
const assert = require('assert');

describe('test/mock.test.js', () => {
  let app;
  before(async function() {
    app = mm.app({
      baseDir: 'apps/mock',
    });
    await app.ready();
  });
  afterEach(mm.restore);
  after(async function() {
    await app.close();
  });

  it('should app.mockProxy ok', async function() {
    app.mockProxy('DemoService', 'sayHello', async function(name) {
      await sleep(1000);

      return 'hello ' + name + ' from mock';
    });

    const ctx = app.createAnonymousContext();
    const res = await ctx.proxy.demoService.sayHello('gxcsoccer');
    assert(res === 'hello gxcsoccer from mock');
  });
});
```

RPC Server Unittest

```js
'use strict';

const mm = require('egg-mock');

describe('test/index.test.js', () => {
  let app;
  before(async function() {
    app = mm.app({
      baseDir: 'apps/rpcserver',
    });
    await app.ready();
  });
  after(async function() {
    await app.close();
  });

  it('should invoke HelloService', done => {
    app.rpcRequest('org.eggjs.rpc.test.HelloService')
      .invoke('hello')
      .send([ 'gxcsoccer' ])
      .expect('hello gxcsoccer', done);
  });
});
```

For more details of `app.rpcRequest`, you can refer to this [acticle](https://github.com/eggjs/egg-sofa-rpc/wiki/%E5%8D%95%E5%85%83%E6%B5%8B%E8%AF%95-RPC-%E6%9C%8D%E5%8A%A1%E7%9A%84%E6%96%B9%E6%B3%95)

## Docs

- [RPC in Nodejs Part One](https://github.com/alipay/sofa-rpc-node/wiki/%E8%81%8A%E8%81%8A-Nodejs-RPC%EF%BC%88%E4%B8%80%EF%BC%89)
- [Cross-Language Interoperability between Eggjs & SOFA](https://github.com/eggjs/egg-sofa-rpc/wiki/Eggjs-%E5%92%8C-SOFA-%E7%9A%84%E8%B7%A8%E8%AF%AD%E8%A8%80%E4%BA%92%E8%B0%83)
- [Custom Service Discovery in Eggjs](https://github.com/eggjs/egg-sofa-rpc/wiki/%E8%87%AA%E5%AE%9A%E4%B9%89%E6%9C%8D%E5%8A%A1%E5%8F%91%E7%8E%B0%E5%AE%9E%E7%8E%B0)
- [RPC Proxy Configuration in Eggjs](https://github.com/eggjs/egg-sofa-rpc/wiki/RPC-%E4%BB%A3%E7%90%86%EF%BC%88Proxy%EF%BC%89%E9%85%8D%E7%BD%AE)
- [RPC Unittest in Eggjs](https://github.com/eggjs/egg-sofa-rpc/wiki/%E5%8D%95%E5%85%83%E6%B5%8B%E8%AF%95-RPC-%E6%9C%8D%E5%8A%A1%E7%9A%84%E6%96%B9%E6%B3%95)

## How to Contribute

Please let us know how can we help. Do check out [issues](https://github.com/eggjs/egg/issues) for bug reports or suggestions first.

To become a contributor, please follow our [contributing guide](https://github.com/eggjs/egg/blob/master/CONTRIBUTING.md).

## License

[MIT](LICENSE)


History.md

1.3.1 / 2019-01-17
==================

**fixes**
  * [[`3d8b65e`](http://github.com/eggjs/egg-rpc/commit/3d8b65e4856bfd7049bf4262649e9c2e95f8ca16)] - fix: add apiMeta to rpc server (gxcsoccer <<gxcsoccer@126.com>>)

1.3.0 / 2018-11-07
==================

**features**
  * [[`fe76cd5`](http://github.com/eggjs/egg-rpc/commit/fe76cd58ad08a8defd3ac1c8082b84391c5e6f40)] - feat: registry class support app (leoner <<kangpangpang@gmail.com>>)

1.2.0 / 2018-11-06
==================

**features**
  * [[`065dd45`](http://github.com/eggjs/egg-rpc/commit/065dd45647f4825c9c7ec5d391215d8e7c94643c)] - feat: add mockProxy api (gxcsoccer <<gxcsoccer@126.com>>)

1.1.0 / 2018-11-03
==================

**features**
  * [[`21f8b90`](http://github.com/eggjs/egg-rpc/commit/21f8b904fd3c72394b6ff81bba6da454e19b0d5b)] - feat: auto load proxy class (gxcsoccer <<gxcsoccer@126.com>>)

1.0.0 / 2018-11-02
==================

**others**
,fatal: 没有发现名称，无法描述任何东西。



config/config.default.js
'use strict';

const fs = require('fs');
const path = require('path');
const antpb = require('antpb');
const protocol = require('sofa-bolt-node');
const { ZookeeperRegistry } = require('sofa-rpc-node').registry;

module.exports = appInfo => {
  let proto;
  let classMap;
  let apiMeta;
  const protoPath = path.join(appInfo.baseDir, 'run/proto.json');
  const proxyClassDir = path.join(appInfo.baseDir, 'app/proxy_class');
  const apiMetaPath = path.join(appInfo.baseDir, 'config/apiMeta.json');
  // 加载 proto
  if (fs.existsSync(protoPath)) {
    proto = antpb.fromJSON(require(protoPath));
  }
  // 加载 classMap
  if (fs.existsSync(proxyClassDir)) {
    classMap = new Proxy({}, {
      get(target, className) {
        let map = target[className];
        if (!map) {
          const args = className.split('.');
          args.unshift(proxyClassDir);
          args[args.length - 1] = args[args.length - 1] + '.js';
          const classfile = path.join.apply(null, args);
          if (fs.existsSync(classfile)) {
            map = target[className] = require(classfile)[className];
          }
        }
        return map;
      },
    });
  }
  protocol.setOptions({ proto, classMap });

  if (fs.existsSync(apiMetaPath)) {
    apiMeta = require(apiMetaPath);
  }
  return {
    rpc: {
      registryClass: ZookeeperRegistry,
      registry: null,
      client: {
        protocol,
        responseTimeout: 3000,
      },
      server: {
        apiMeta,
        protocol,
        port: 12200,
        idleTime: 5000,
        killTimeout: 30000,
        maxIdleTime: 90 * 1000,
        responseTimeout: 3000,
        codecType: 'protobuf',
        selfPublish: true,
        // 下面配置针对新的 rpc 服务发布方式
        namespace: null,
        version: '1.0',
        group: 'SOFA',
        uniqueId: null,
        autoServe: true, // 如果发现有暴露服务，则自定启动 server
      },
    },
  };
};
